x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/RollupShardPersistentTaskExecutor.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.downsample;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.index.IndexService;
+import org.elasticsearch.index.mapper.TimeSeriesIdFieldMapper;
+import org.elasticsearch.index.query.MatchAllQueryBuilder;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.persistent.AllocatedPersistentTask;
+import org.elasticsearch.persistent.PersistentTaskState;
+import org.elasticsearch.persistent.PersistentTasksCustomMetadata;
+import org.elasticsearch.persistent.PersistentTasksExecutor;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.sort.SortOrder;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.xpack.core.rollup.action.RollupShardIndexerStatus;
+import org.elasticsearch.xpack.core.rollup.action.RollupShardPersistentTaskState;
+import org.elasticsearch.xpack.core.rollup.action.RollupShardTask;
+import org.elasticsearch.xpack.rollup.Rollup;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+
+public class RollupShardPersistentTaskExecutor extends PersistentTasksExecutor<RollupShardTaskParams> {
+    private final Client client;
+    private final IndicesService indicesService;
+
+    public RollupShardPersistentTaskExecutor(
+        final Client client,
+        final IndicesService indicesService,
+        final String taskName,
+        final String executorName
+    ) {
+        super(taskName, executorName);
+        this.client = Objects.requireNonNull(client);
+        this.indicesService = Objects.requireNonNull(indicesService);
+    }
+
+    @Override
+    protected void nodeOperation(final AllocatedPersistentTask task, final RollupShardTaskParams params, final PersistentTaskState state) {
+        // NOTE: query the downsampling target index so that we can start the downsampling task from the latest indexed tsid.
+        final SearchHit[] lastRollupTsidHits = client.prepareSearch(params.rollupIndex())
+            .addSort(TimeSeriesIdFieldMapper.NAME, SortOrder.DESC)
+            .setSize(1)
+            .setQuery(new MatchAllQueryBuilder()) // TODO: can we do better then querying all documents?
```

Comment:
I think we need to set search preference to `_shard: shardID`, otherwise we query all shards or the target index. And we get tsids that this `RollupShardIndexer` never downsampled. This should work because the target index always has the same number of shards of the source index. So tsids in source index shard 2 will be indexed in target index shard 2.